### PR TITLE
[BUILD] Add support for bzlmod

### PR DESCRIPTION
**Pull request by keith**
This adds support for bzlmod, which is bazel's new dependency resolution system.

Theoretically we could now remove the previous dependency management configuration which required flattening all the dependencies, but I left it here for now for users using old versions of bazel.

Fixes: https://github.com/open-telemetry/opentelemetry-cpp/issues/2334

### CODE REVIEW

**Comment by keith** [comment]
ok I pushed an update here that seems to fix it at least with c++17, it depends on https://github.com/bazelbuild/bazel-central-registry/pull/1810

**Comment by marcalff** [comment]
Hi @keith 

Could you merge with a recent main to resolve conflicts,
and investigate the remaining CI failures for bazel if any ?

Thanks,
-- Marc

**Comment by keith** [comment]
The windows and asan failures here are real.

**Comment by keith** [comment]
hopefully will be unblocked by https://github.com/open-telemetry/opentelemetry-cpp/pulls/keith

**Comment by marcalff** [comment]
@keith 

Now that #2679 is merged, could you update this PR and resolve conflicts ?

Looking to include this in the upcoming release.

Thanks.

**Comment by keith** [comment]
I think there might still be a grpc ODR issue I have to fix, but we'll see if CI fails

File: MODULE.bazel [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Copyright The OpenTelemetry Authors
+# SPDX-License-Identifier: Apache-2.0
+
+module(
+    name = "opentelemetry-cpp",
+    version = "1.14.2",
```
Comment (marcalff) [inline MODULE.bazel:6]:
```suggestion
    version = "1.15.0",
```

I assume this should match the current version in the opentelemetry-cpp repository.

File: MODULE.bazel [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    compatibility_level = 0,
+    repo_name = "io_opentelemetry_cpp",
+)
+
+bazel_dep(name = "abseil-cpp", version = "20240116.1", repo_name = "com_google_absl")
+bazel_dep(name = "bazel_skylib", version = "1.5.0")
+bazel_dep(name = "curl", version = "8.4.0")
+bazel_dep(name = "grpc", version = "1.62.1", repo_name = "com_github_grpc_grpc")
+bazel_dep(name = "nlohmann_json", version = "3.11.3", repo_name = "github_nlohmann_json")
+bazel_dep(name = "opentelemetry-proto", version = "1.1.0", repo_name = "com_github_opentelemetry_proto")
```
Comment (marcalff) [inline MODULE.bazel:16]:
```suggestion
bazel_dep(name = "opentelemetry-proto", version = "1.3.1", repo_name = "com_github_opentelemetry_proto")
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -537,6 +537,27 @@ jobs:
     - name: run tests
       run: ./ci/do_ci.sh bazel.test
 
+  bazel_no_bzlmod_test:
+    name: Bazel
```
Comment (marcalff) [inline .github/workflows/ci.yml:541]: There is already a CI build named Bazel, please pick something more descriptive.

File: ci/do_ci.sh [hunk 4]
```diff
@@ -459,6 +459,10 @@ EOF
   make load_plugin_example
   examples/plugin/load/load_plugin_example ${PLUGIN_DIR}/libexample_plugin.so /dev/null
   exit 0
+elif [[ "$1" == "bazel.no_bzlmod.test" ]]; then
+  bazel $BAZEL_STARTUP_OPTIONS build --enable_bzlmod=false $BAZEL_OPTIONS //...
+  bazel $BAZEL_STARTUP_OPTIONS test --enable_bzlmod=false $BAZEL_TEST_OPTIONS //...
```
Comment (marcalff) [inline ci/do_ci.sh:464]: So, does this build ignores the content of file MODULE.bazel ?

Should we have another build with bzlmod enabled ?

**Review by marcalff — COMMENTED** [review]
Thanks for the PR.

See minor fixes, and some questions.

File: ci/do_ci.sh (hunk 4 above)
Comment (keith) [reply ci/do_ci.sh:464]: bazel 7.x enables bzlmod by default, so when this merges all bazel jobs that already exist will use this new method by default. so I added this one to make sure the old way continues to work at least for this single build, and figure we can run all the other CI jobs with the new way

File: MODULE.bazel (hunk 1 above)
Comment (keith) [reply MODULE.bazel]: actually this number just needs to be replaced when it's uploaded to the BCR with a patch, so I actually just zero'd this out

File: MODULE.bazel (hunk 2 above)
Comment (keith) [reply MODULE.bazel:16]: looks like this version isn't in the BCR yet, so I'll have to add that there first

Comment (keith) [reply MODULE.bazel:16]: https://github.com/bazelbuild/bazel-central-registry/pull/2252

Comment (marcalff) [reply MODULE.bazel:16]: proto 1.3.1 is now merged in BCR, upgrading to 1.3.1

**Comment by keith** [comment]
thanks!

**Pull request merged**